config.yaml

```diff
@@ -59,3 +59,15 @@ options:
       This configuration accepts comma-separated MAC prefixes in string format.
     default: "52:54:00,fa:16:3e,06:f1:3a,00:0d:3a,00:50:56"
     type: string
+  skip-interfaces:
+    description: |
+      The list of interface names that should not be considered when detecting machine
+      type. This is useful, for example, when a machine is a KVM host (not guest) and it
+      has a bridge (e.g. "virbr0") with mac address matching one of the prefixes
+      in "virtual-macs" option. Such interface would then cause this machine to be mislabeled
+      as "KVM" type.
+      This configuration accepts comma-separated regular expressions in string format.
+      Any interface that matches any of these expressions will not be considered when
+      detecting machine type.
+    default: "^tap-,^vribr\\d*,^lxdbr\\d*,^docker\\d*,^fan-\\d*,^vnet\\d*,^vxlan,^genev,^(tun|tap|cali|veth)[a-f0-9]*"
```

Comment:
so we're allowing users to specify lists of regexps but... what if they want ',' to be part of the regex?
could we perhaps have this be a json list instead?
it would be nice to also be able to avoid escaping backslashes, that was unexpected

Reply:
Yeah, that's unfortunately a limitation of this approach but is there a realistic need to write complex regexps with comas? Wouldn't most of the input be done in the form of "starts-with" and "contains"?

> it would be nice to also be able to avoid escaping backslashes, that was unexpected

So, I tested this and it actually have to be done like this only in the `config.yaml`. They appear without double backslashes in the output of `juju config` and they don't need to be "double escaped" in the input for `juju config` either.

Reply:
That being said, we can do the "json list" approach too. I can get to it first thing on Monday. When it's fixed, is this PR ready to land from your perspective?

Reply:
Yes, I suspect that 99.9% of the time regexps will be overkill, and that substrings would be sufficient. Also yes, aside from the usability concern, I am +1 on this change.
It's great that as a user you don't need to type `\\d` in `juju config`, so I'd suggest either:
1. switching to json if it's not too much of a hassle; it would also have the slight added benefit of aligning with already-existing config options such as nova-compute's sysctl and pci-passthrough-whitelist
2. combining the list of regexps into a single regexp, essentially by replacing commas with pipes. I don't think it would significantly worsen the experience. The expression would look like this (after fixing a typo on virbr and adding macvtap): `^((virbr|lxdbr|docker|fan-|vnet|macvtap)\d*|vxlan|genev|(tun|tap|cali|veth)[a-f0-9]*)`

Reply:
So I thought about this and tested bit more and:

- Approach `#1` requires double escaping `\` otherwise json refuses to load it (I think it's the same reason why it must be double-escaped in `config.yaml`). So that'd be a downgrade in user experience imo.
- The approach `#2` is already possible. User can type just one regexp covering all possible matches. However, changing to single regexp exclusively would require change in the snap as well (changing config option from list of strings to single string.).
- After looking through few documentations and guides, [comma seems to be used only in range quantifiers](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Cheatsheet#quantifiers) and have no other special meaning.

Give  these factors, I'd say we should stick with the original comma separated strings.